pkgs/desktops/gnome-3/extensions/update-extensions.py

```diff
@@ -0,0 +1,247 @@
+# Group extensions by Gnome "major" version for compatibility reasons
+supported_versions = {
+    "36": "3.36",
+    "38": "3.38",
+}
+
+
+def fetch_sha256sum(uuid: str, version: str) -> str:
+    """
+    Download the extension and hash it. We use `nix-prefetch-url` for this for efficiency reasons
+    """
+    import subprocess
+
+    # The download URLs follow this schema
+    uuid = uuid.replace("@", "")
+    url: str = f"https://extensions.gnome.org/extension-data/{uuid}.v{version}.shell-extension.zip"
+
+    return subprocess.run(
+        ["nix-prefetch-url", url], capture_output=True, text=True
+    ).stdout.strip()
+
+
+def generate_extension_versions(
+    extension_version_map: Dict[str, str], uuid: str
+) -> Dict[str, Dict[str, str]]:
+    """
+    Takes in a mapping from shell versions to extension versions and transforms it the way we need it:
+    - Only take one extension version per Gnome Shell major version (as per `supported_versions`)
+    - Filter out versions that support old Gnome versions
+    - Download the extension and hash it
+    """
+    extension_versions: Dict[str, Dict[str, str]] = {}
+    for shell_version, version_prefix in supported_versions.items():
+        # Newest compatible extension version
+        extension_version: Optional[int] = max(
+            (
+                int(ext_ver)
+                for shell_ver, ext_ver in extension_version_map.items()
+                if (shell_ver.startswith(version_prefix))
+            ),
+            default=None,
+        )
+        # Extension is not compatible with this Gnome version
+        if not extension_version:
+            continue
+        logging.debug(
+            f"[{shell_version}] Downloading '{uuid}' v{str(extension_version)}"
+        )
+        sha256 = fetch_sha256sum(uuid, str(extension_version))
+        extension_versions[shell_version] = {
+            "version": str(extension_version),
+            "sha256": sha256,
+        }
+    return extension_versions
+
+
+def pname_from_url(url: str) -> Tuple[str, str]:
+    """
+    Parse something like "/extension/1475/battery-time/" and output ("battery-time", "1475")
+    """
+
+    url = url.split("/")  # type: ignore
+    return (url[3], url[2])
+
+
+def sanitize_string(string: str) -> str:
+    """
+    Replace thinkgs like "\\u0123" in strings. This is required for nix < 2.3.10, as it does not support "\\u"
+    """
+    # The encode-decode trick also produces "\\n", but Nix supports "\n".
+    return string.encode("unicode-escape").decode().replace("\\n", "\n")
+
+
+# Keep track of all names that have been used till now to detect collisions.
+# This works because we deterministically process all extensions in historical order
+# The outer dict level is the shell version, as we are tracking duplicates only per same Shell version.
+# key: shell version, value: Dict with key: pname, value: list of UUIDs with that pname
+package_name_registry: Dict[str, Dict[str, List[str]]] = {}
```

Comment:
Please keep the code that does "static initialisation" to one place. Can you move this up to the place where you define the supported versions?